dotnet/src/SemanticKernel/Planning/SKContextExtensions.cs

```diff
@@ -2,47 +2,162 @@
-    internal static List<FunctionView> GetAvailableFunctions(
+    /// <summary>
+    /// Returns a list of functions that are available to the user based on the semantic query and the excluded skills and functions.
+    /// </summary>
+    /// <param name="context">The SKContext</param>
+    /// <param name="config">The planner skill config.</param>
+    /// <param name="semanticQuery">The semantic query for finding relevant registered functions</param>
+    /// <returns>A list of functions that are available to the user based on the semantic query and the excluded skills and functions.</returns>
+    internal static async Task<List<FunctionView>?> GetAvailableFunctionsAsync(
         this SKContext context,
-        List<string>? excludedSkills = null,
-        List<string>? excludedFunctions = null)
+        PlannerSkillConfig config,
+        string? semanticQuery = null)
     {
-        excludedSkills ??= new();
-        excludedFunctions ??= new();
+        var excludedSkills = config.ExcludedSkills ?? new();
+        var excludedFunctions = config.ExcludedFunctions ?? new();
+        var includedFunctions = config.IncludedFunctions ?? new();
 
         context.ThrowIfSkillCollectionNotSet();
 
         var functionsView = context.Skills!.GetFunctionsView();
 
-        return functionsView.SemanticFunctions
+        var availableFunctions = functionsView.SemanticFunctions
             .Concat(functionsView.NativeFunctions)
             .SelectMany(x => x.Value)
             .Where(s => !excludedSkills.Contains(s.SkillName) && !excludedFunctions.Contains(s.Name))
             .ToList();
+
+        List<FunctionView> result = availableFunctions;
+        if (!string.IsNullOrEmpty(semanticQuery))
+        {
+            // If a Memory provider has not been registered, do not filter the functions.
+            if (context.Memory is not NullMemory)
+            {
+                // catalog functions to memory
+                foreach (var function in availableFunctions)
+                {
+                    var functionName = function.ToFunctionName();
+                    var key = string.IsNullOrEmpty(function.Description) ? functionName : function.Description;
+
+                    // It'd be nice if there were a saveIfNotExists method on the memory interface
+                    var memoryEntry = await context.Memory.GetAsync(MemoryCollectionName, key, context.CancellationToken);
+                    if (memoryEntry == null)
+                    {
+                        // TODO It'd be nice if the minRelevanceScore could be a parameter for each item that was saved to memory
+                        // As folks may want to tune their functions to be more or less relevant.
+                        await context.Memory.SaveInformationAsync(MemoryCollectionName, key, functionName, function.ToManualString(),
+                            context.CancellationToken);
+                    }
+                }
```

Comment:
I feel like this whole chunk (// catalog functions to memory...) could at minimum be its own function. Also questioning whether this is the right place to be doing this. Does GetAvailableFunctionsAsync get called repeatedly?

Reply:
Good call. Could see argument to make this part of import or similar, too. Yeah, it does get called every time a plan is created.

Reply:
For now, I will refactor into its own method at least, would like to address doing this in the import path with @dluc separately as this would make our kernel Register/Import entry points asynchronous.